format/postgres/common/checksum.go

```diff
@@ -0,0 +1,68 @@
+package common
+
+import "encoding/binary"
+
+const (
+	nSums        = 32
+	fnvPrime     = 16777619
+	nSumsSize    = 4 * nSums
+	mainBlockLen = PageSize / nSumsSize
+	RelSegSize   = 131072
+)
+
+var (
+	checksumBaseOffsets = [nSums]uint32{
+		0x5B1F36E9, 0xB8525960, 0x02AB50AA, 0x1DE66D2A,
+		0x79FF467A, 0x9BB9F8A3, 0x217E7CD2, 0x83E13D2C,
+		0xF8D4474F, 0xE39EB970, 0x42C6AE16, 0x993216FA,
+		0x7B093B5D, 0x98DAFF3C, 0xF718902A, 0x0B1C9CDB,
+		0xE58F764B, 0x187636BC, 0x5D7B3BB1, 0xE73DE7DE,
+		0x92BEC979, 0xCCA6C0B2, 0x304A0979, 0x85AA43D4,
+		0x783125BB, 0x6CA8EAA2, 0xE407EAC6, 0x4B5CFC3E,
+		0x9FBF8C76, 0x15CA20BE, 0xF2CA9FD3, 0x959BD756,
```

Comment:
Is this a well known checksum or postgres specific? The go standard library has some standard ones and fq has a checksum package with some additional ones. They all implement the hash.Hash interface. Had some plan for a while to make the decode API be able to do something generic for checksums but got complicated, maybe in the future. One things nice with hash.Hash interface design is that it embeds a io.Writer so they are meant to use used with io.Copy etc (see flac_frame)

Reply:
It's looks like FNV-1a. But PostgrerSQL hashes doesn't match to FNV-1a hashes in Golang.
This hash is PostgreSQL specific.

Reply:
Ok, would it be possible implement the postgres variant based on some of the fvn code in the standard library (copy the code and change it)? Thinking it would be nice to have it as a hash.Hash implementation to make it more similar to checksumming in other decoders and also prepare for more generic checksum support in the decode API.
I've thought about doing something similar to what you do now, having a struct with calculated, actual and valid fields etc and it would be used some like this: `d.FieldChecksum("page_checksum", start, stop, [zeroGaps somehow?], crc32.New())` and it would take care of add a struct, fields, do all the io stuff etc.